main/src/ca/uwaterloo/flix/language/phase/Lowering.scala

```diff
@@ -472,24 +477,192 @@ object Lowering extends Phase[Root, Root] {
+      val sym = Symbol.mkDefnSym("Channel.put")
+      val callExp = Expression.Def(sym, defTpe, loc)
+      val args = List(e1, e2)
+      Expression.Apply(callExp, args, chanType, eff, loc)
+
+    // Translate this expression to:
+    // ```
+    // match selectImpl([ch1, ch2], false) {
+    //   case None => e3 (or into MatchError if there is no default case).
+    //   case Some(res) => match res {
+    //     case (0, IntResult(x)) => e1
+    //     case (1, AnyResult(x')) => let x = x' as String; e2
+    //     case _ => MatchError
+    //   }
+    // }
+    // ```
     case Expression.SelectChannel(rules, default, tpe, eff, loc) =>
       val rs = rules.map(visitSelectChannelRule)
       val d = default.map(visitExp)
       val t = visitType(tpe)
-      Expression.SelectChannel(rs, d, t, eff, loc)
+
+      // Call `selectImpl([<scala> rs.map(rule => rule.chan) </scala>],
+      //    <scala> d match { case Some(_) => Expression.True(loc), case None => Expression.False(loc) } </scala>,)`
+      val indexElementTupleType = Type.mkTuple(List(Type.mkInt32(loc), Types.SelectResult))
+      val optionSym = Symbol.mkEnumSym("Option")
+      val selectImplReturnTpe = Type.mkEnum(optionSym, List(indexElementTupleType))
+      val selectImplDefTpe = Type.mkImpureUncurriedArrow(List(Type.mkArray(Types.ChannelImpl, loc), Type.mkBool(loc)), selectImplReturnTpe)
+      val selectImplSym = Symbol.mkDefnSym("Channel.selectImpl")
+      val selectImplCallExp = Expression.Def(selectImplSym, selectImplDefTpe, loc)
+      val channelsInRs = rs.map(rule => rule.chan)
+      val hasDefault = d match {
+        case Some(_) => Expression.True(loc)
+        case None => Expression.False(loc)
+      }
+      val selectImplArgs = List(Expression.ArrayLit(channelsInRs, Types.ChannelImpl, Type.Impure, loc), hasDefault)
+      val selectImplApply = Expression.Apply(selectImplCallExp, selectImplArgs, selectImplReturnTpe, eff, loc)
+
+      // The code below should implement this
+      // ```
+      // match <scala> selectImplApply </scala> {
+      //  case None => e3 (or into MatchError if there is no default case).
+      //  case Some(res) => match res {
+      //    case (0, IntResult(x)) => e1
+      //    case (1, AnyResult(x')) => let x = x' as String; e2
+      //    case _ => MatchError
+      //  }
+      // }
+      // ```
+
+      val sourcePositionBegin = SourcePosition(loc.source, loc.beginLine, loc.beginCol, None)
+      val sourcePositionEnd = SourcePosition(loc.source, loc.endLine, loc.endCol, None)
+
+      // None case
+      val noneTag = Name.mkTag(Name.Ident(sourcePositionBegin, "None", sourcePositionEnd))
+      val nonePattern = Pattern.Tag(optionSym, noneTag, Pattern.Unit(loc), t, loc)
+      val noneMatchRule = MatchRule(nonePattern, Expression.True(loc), d match {
+        case Some(e) => e
+        case None => ??? // TODO: 1 - Q: How to return a `MatchError` on the `TypedAST`?
```

Comment:
> Do you know how to go about this? I think MatchError only become available on the SimplifiedAST. Does that mean I should just add a MatchError case to TypedAST or am I missing something?

@magnus-madsen, just re-commenting on this since the last comment became outdated. 

Do you have thoughts on how to do this? Am I misunderstanding you when trying to return a `MatchError` `case class` of `TypedAST`?